packages/arch/packages/tooltip/src/index.js

```diff
@@ -57,118 +56,122 @@ const showTooltip = (fn, defaultDelay) => {
   if (isHidePending) {
     pendingHide.flush();
   }
-  const pendingShow = flushable(() => fn(isHidePending), isHidePending ? 0 : defaultDelay);
+
+  const pendingShow = flushable(fn, isHidePending ? 0 : defaultDelay);
   return pendingShow.cancel;
 };
 
 const hideTooltip = (fn, defaultDelay) => {
-  pendingHide = flushable(flushed => fn(flushed), defaultDelay);
+  pendingHide = flushable(fn, defaultDelay);
   return pendingHide.cancel;
 };
 
-export default class Tooltip extends Component {
-  state = {
-    immediatelyHide: false,
-    immediatelyShow: false,
-    isVisible: false,
-  };
-  ref = createRef();
-  cancelPendingSetState = NOOP;
-  static defaultProps = {
-    delay: 300,
-    placement: 'bottom',
-  };
-  componentDidMount() {
-    const target = this.ref.current;
+const Tooltip = ({
+  children,
+  content,
+  onHide,
+  onShow,
+  placement,
+  className,
+  hideOnMouseDown,
+  hideOnKeyDown,
+  delay,
+}) => {
+  const [isVisible, setIsVisible] = useState(false);
+  const ref = useRef();
+
+  useEffect(() => {
+    const target = ref.current;
 
     if (!target) {
       throw new Error('You must pass the ref onto your target node.');
     }
+
     if (!target.nodeName) {
       throw new Error(
         "It looks like you've passed the ref onto a component. You must pass the ref onto your target node."
       );
     }
 
-    target.addEventListener('mouseenter', this.handleMouseEnter, LISTENER_OPTIONS);
-    target.addEventListener('mouseleave', this.handleMouseLeave, LISTENER_OPTIONS);
-  }
-  componentWillUnmount() {
-    this.cancelPendingSetState();
-
-    const target = this.ref.current;
-
-    if (target) {
-      target.removeEventListener('mouseenter', this.handleMouseEnter, LISTENER_OPTIONS);
-      target.removeEventListener('mouseleave', this.handleMouseLeave, LISTENER_OPTIONS);
-    }
-  }
-
-  cancel = () => {
-    this.cancelPendingSetState();
-    this.setState({ isVisible: false, immediatelyHide: true });
-  };
-
-  handleMouseEnter = () => {
-    this.cancelPendingSetState();
-
-    if (this.state.isVisible) {
-      return;
-    }
-    if (this.props.hideOnMouseDown && this.ref.current) {
-      this.ref.current.addEventListener('mousedown', this.cancel, LISTENER_OPTIONS);
-    }
-    if (this.props.hideOnKeyDown) {
-      document.addEventListener('keydown', this.cancel, LISTENER_OPTIONS);
-    }
+    let cancelPendingSetState = () => {};
+
+    const cancel = () => {
```

Comment:
It may be a bit more succinct/readable to pull these event handlers outside of the `useEffect` and limit the body of the `useEffect` to registering + removing the event listeners.

Reply:
I could move `cancelPendingSetState` out, but everything else uses props/state.

Reply:
Sorry, to clarify, I didn't mean outside of the component, but outside of the `useEffect` method.

I think `cancel`, `handleMouseEnter` and `handleMouseLeave` should all live above the `useEffect` method.

Reply:
I think that makes things just a little messier since then `cancelPendingSetState` has to be managed with `useRef` as well. 🤔